acceptancetests/jujupy/k8s_provider/gke.py

```diff
@@ -0,0 +1,283 @@
+        cfg = env._config
+        self.default_params = dict(
+            project_id=cfg['project_id'],
+            zone=zone,
+        )
+
+        self.driver = container_v1.ClusterManagerClient(
+            credentials=service_account.Credentials.from_service_account_info(cfg)
+        )
+        
+        self.__ensure_gcloud(cfg)
+
+        # list all running clusters
+        running_clusters = self.driver.list_clusters(**self.default_params)
+        logger.info('running gke clusters: %s', running_clusters)
+
+    def _ensure_cluster_stack(self):
+        self.provision_gke()
+        
+    def __ensure_gcloud(self, cfg):
+        gcloud_bin = shutil.which('gcloud')
+        if gcloud_bin is None:
+            # ensure gcloud installed.
+            self.sh('sudo', 'snap', 'install', 'google-cloud-sdk', '--classic')
+            logger.debug("gcloud installed successfully")
+            gcloud_bin = shutil.which('gcloud')
+        
+        cred = {
+            k: cfg[k] for k in (
+                'project_id', 
+                'private_key_id', 
+                'private_key', 
+                'client_email', 
+                'client_id', 
+                'auth_uri', 
+                'token_uri', 
+                'auth_provider_x509_cert_url', 
+                'client_x509_cert_url',
+            )
+        }
+        cred['type'] = 'service_account'
+        cred_file = tempfile.NamedTemporaryFile(suffix='.json', delete=False)
+        try:
+            with open(cred_file.name, 'w') as f:
+                json.dump(cred, f)
+                
+            # set credential;
+            o = self.sh(gcloud_bin, 'auth', 'activate-service-account', '--key-file', cred_file.name)
+            logger.debug(o)
+            # set project;
+            o = self.sh(gcloud_bin, 'config', 'set', 'project', cfg['project_id'])
+            logger.debug(o)
+        finally:
+            os.unlink(cred_file.name)
+
+    def _tear_down_substrate(self):
+        try:
+            self.driver.delete_cluster(
+                cluster_id=self.gke_cluster_name, **self.default_params,
+            )
+        except exceptions.NotFound:
+            ...
+
+    def _ensure_kube_dir(self):
+        cluster = self._get_cluster(self.gke_cluster_name)
+        cluster_config = ClusterConfig(
+            project_id=self.default_params['project_id'],
+            cluster=cluster,
+        )
+        self.kubeconfig_cluster_name = cluster_config.context_name
+        kubeconfig_content = cluster_config.dump()
+        with open(self.kube_config_path, 'w') as f:
+            logger.debug('writing kubeconfig to %s\n%s', self.kube_config_path, kubeconfig_content)
+            f.write(kubeconfig_content)
+
+        # ensure kubectl
+        kubectl_bin_path = shutil.which('kubectl')
+        if kubectl_bin_path is not None:
+            self.kubectl_path = kubectl_bin_path
+        else:
+            self.sh(
```

Comment:
I don't have kubectl installed and it doesn't look like this fallback worked - I get:
```
2019-06-26 14:07:56 ERROR [Errno 2] No such file or directory: '/tmp/test-run/juju-homes/nw-caas-deploy-charms-gke/kubectl': '/tmp/test-run/juju-homes/nw-caas-deploy-charms-gke/kubectl'
Traceback (most recent call last):
  File "/home/xtian/dev/go/src/github.com/juju/juju/acceptancetests/utility.py", line 416, in logged_exception
    yield
  File "/home/xtian/dev/go/src/github.com/juju/juju/acceptancetests/deploy_stack.py", line 1034, in runtime_context
    yield
  File "/home/xtian/dev/go/src/github.com/juju/juju/acceptancetests/deploy_stack.py", line 1177, in existing_booted_context
    yield machines
  File "./acceptancetests/assess_caas_deploy_charms.py", line 174, in main
    assess_caas_charm_deployment(caas_client, args.caas_provider)
  File "./acceptancetests/assess_caas_deploy_charms.py", line 105, in assess_caas_charm_deployment
    if not caas_client.check_cluster_healthy(timeout=60):
  File "/home/xtian/dev/go/src/github.com/juju/juju/acceptancetests/jujupy/k8s_provider/base.py", line 194, in check_cluster_healthy
    if check():
  File "/home/xtian/dev/go/src/github.com/juju/juju/acceptancetests/jujupy/k8s_provider/base.py", line 185, in check
    cluster_info = self.kubectl('cluster-info')
  File "/home/xtian/dev/go/src/github.com/juju/juju/acceptancetests/jujupy/k8s_provider/base.py", line 200, in kubectl
    return self.sh(*(self._kubectl_bin + args))
  File "/home/xtian/dev/go/src/github.com/juju/juju/acceptancetests/jujupy/k8s_provider/base.py", line 217, in sh
    stderr=subprocess.STDOUT,
  File "/usr/lib/python3.7/subprocess.py", line 395, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.7/subprocess.py", line 472, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/test-run/juju-homes/nw-caas-deploy-charms-gke/kubectl': '/tmp/test-run/juju-homes/nw-caas-deploy-charms-gke/kubectl'
```

Reply:
Rerunning it now that you've changed this bit.